gossip/emitter/throttler/dominant_set_test.go

```diff
@@ -0,0 +1,245 @@
+// (at your option) any later version.
+//
+// Sonic is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with Sonic. If not, see <http://www.gnu.org/licenses/>.
+
+package throttler
+
+import (
+	"fmt"
+	"maps"
+	"math"
+	"math/rand"
+	"slices"
+	"testing"
+
+	"github.com/Fantom-foundation/lachesis-base/inter/idx"
+	"github.com/Fantom-foundation/lachesis-base/inter/pos"
+	"github.com/stretchr/testify/require"
+)
+
+func TestComputeDominantSet_CanIdentifyWhenStakeDistributionIsDominated(t *testing.T) {
+
+	// This test covers various basic cases for ease of development
+
+	const testThreshold = 0.75
+
+	tests := map[string]struct {
+		stakes      []int64
+		expectedSet []idx.ValidatorID
+	}{
+		"no validators": {
+			stakes: nil,
+		},
+		"single validator": {
+			stakes:      []int64{100},
+			expectedSet: []idx.ValidatorID{1},
+		},
+		"two equal validators": {
+			stakes:      []int64{50, 50},
+			expectedSet: []idx.ValidatorID{1, 2},
+		},
+		"two validators one dominant": {
+			stakes:      []int64{80, 20},
+			expectedSet: []idx.ValidatorID{1},
+		},
+		"three validators one dominant": {
+			stakes:      []int64{80, 10, 10},
+			expectedSet: []idx.ValidatorID{1},
+		},
+		"three validators two dominant": {
+			stakes:      []int64{40, 40, 20},
+			expectedSet: []idx.ValidatorID{1, 2},
+		},
+		"four equal validators, first three dominate": {
+			stakes:      []int64{25, 25, 25, 25},
+			expectedSet: []idx.ValidatorID{1, 2, 3},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			stakes := makeValidatorsFromStakes(test.stakes...)
+			set := ComputeDominantSet(stakes, pos.Weight(100), testThreshold)
+			require.ElementsMatch(t, test.expectedSet, slices.Collect(maps.Keys(set)))
+		})
+	}
+}
+
+func TestComputeDominantSet_UniformDistributionsAreDominated(t *testing.T) {
+	// A set of n validators with equal stake shall have a dominant set
+	// of size ceil(n * threshold)
+
+	for threshold := float64(0.67); threshold <= 1.; threshold += 0.05 {
+		for n := 1; n <= 100; n++ {
+			t.Run(fmt.Sprintf("n=%d,threshold=%.2f", n, threshold), func(t *testing.T) {
+				validators := makeValidatorsFromStakes(slices.Repeat([]int64{10}, n)...)
+				set := ComputeDominantSet(validators, validators.TotalWeight(), threshold)
+				expectedCount := int(math.Ceil(float64(n) * threshold))
+				require.Len(t, set, expectedCount, "n=%d, threshold=%.2f", n, threshold)
+				require.GreaterOrEqual(t, sumStake(set, validators), validators.TotalWeight()*pos.Weight(threshold))
+			})
+		}
+	}
+}
+
+func TestComputeDominantSet_CanBeEmpty(t *testing.T) {
+
+	t.Run(
+		"nominal stake > than online stake * threshold",
+		func(t *testing.T) {
```

Comment:
purely visual preference.
```suggestion
	t.Run("nominal stake > than online stake * threshold",
		func(t *testing.T) {
```